Mathlib/Data/Ordmap/Ordnode.lean

```diff
@@ -181,254 +188,262 @@ instance {α} [Repr α] : Repr (Ordnode α) :=
-          · exact node 3 (ι x) rx rr
-        · cases' id rr with rrs
-          · exact node 3 (ι x) rlx ι rx
-          ·
-            exact
-              if rls < ratio * rrs then node (rs + 1) (node (rls + 1) nil x rl) rx rr
-              else
-                node (rs + 1) (node (size rll + 1) nil x rll) rlx
-                  (node (size rlr + rrs + 1) rlr rx rr)
-    · cases' id r with rs rl rx rr
       · cases' id ll with lls
-        · cases' lr with _ _ lrx
-          · exact node 2 l x nil
-          · exact node 3 (ι lx) lrx ι x
-        · cases' id lr with lrs lrl lrx lrr
-          · exact node 3 ll lx ι x
-          ·
-            exact
-              if lrs < ratio * lls then node (ls + 1) ll lx (node (lrs + 1) lr x nil)
-              else
-                node (ls + 1) (node (lls + size lrl + 1) ll lx lrl) lrx
-                  (node (size lrr + 1) lrr x nil)
-      · refine'
-          if delta * ls < rs then _ else if delta * rs < ls then _ else node (ls + rs + 1) l x r
-        · cases' id rl with rls rll rlx rlr
-          · exact nil
-          --should not happen
-          cases' id rr with rrs
-          · exact nil
-          --should not happen
-          exact
-            if rls < ratio * rrs then node (ls + rs + 1) (node (ls + rls + 1) l x rl) rx rr
-            else
-              node (ls + rs + 1) (node (ls + size rll + 1) l x rll) rlx
-                (node (size rlr + rrs + 1) rlr rx rr)
-        · cases' id ll with lls
-          · exact nil
-          --should not happen
-          cases' id lr with lrs lrl lrx lrr
-          · exact nil
-          --should not happen
-          exact
-            if lrs < ratio * lls then node (ls + rs + 1) ll lx (node (lrs + rs + 1) lr x r)
-            else
-              node (ls + rs + 1) (node (lls + size lrl + 1) ll lx lrl) lrx
-                (node (size lrr + rs + 1) lrr x r)
+        · exact nil
+        --should not happen
+        cases' id lr with lrs lrl lrx lrr
+        · exact nil
+        --should not happen
+        exact
+          if lrs < ratio * lls then node (ls + rs + 1) ll lx (node (lrs + rs + 1) lr x r)
+          else
+            node (ls + rs + 1) (node (lls + size lrl + 1) ll lx lrl) lrx
+              (node (size lrr + rs + 1) lrr x r)
 #align ordnode.balance Ordnode.balance
 
 /-- O(n). Does every element of the map satisfy property `P`?
 
-     all (λ x, x < 5) {1, 2, 3} = true
-     all (λ x, x < 5) {1, 2, 3, 5} = false -/
+     All (λ x, x < 5) {1, 2, 3} = true
+     All (λ x, x < 5) {1, 2, 3, 5} = false -/
 def All (P : α → Prop) : Ordnode α → Prop
   | nil => True
-  | node _ l x r => all l ∧ P x ∧ all r
+  | node _ l x r => All P l ∧ P x ∧ All P r
 #align ordnode.all Ordnode.All
 
-instance All.decidable {P : α → Prop} [DecidablePred P] (t) : Decidable (All P t) := by
-  induction t <;> dsimp only [all] <;> skip <;> infer_instance
+-- porting notes: required `noncomutable` & can remove `[Decidable P]`
```

Comment:
This is bad. Classically, every proposition is decidable, so this instance is useless if you use `Classical.decPred`.
Re-insert `[DecidablePred P]`, use a combination of `infer_instance` / `All.decidable` (recursively) / manual application of decidability instances (e.g. [And.decidable](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Init/Logic.html#And.decidable)) in the proof, and remove `noncomputable`.

Reply:
I tried to fix this like so:
```lean
instance All.decidable {P : α → Prop} [DecidablePred P] (t) : Decidable (All P t) := by
  match t with
  | nil =>
    dsimp only [All]
    infer_instance
  | node s l x r =>
    have : Decidable (All P r) := by
      exact All.decidable r
    have : Decidable (All P l) := by
      exact All.decidable l
    have hr : Decidable (P x ∧ All P r) := by
      exact And.decidable
    exact And.decidable
    termination_by All.decidable t => t.size
```
but this gives
```
failed to prove termination, possible solutions:
  - Use `have`-expressions to prove the remaining goals
  - Use `termination_by` to specify a different well-founded relation
  - Use `decreasing_by` to specify your own tactic for discharging this kind of goal
```

But even without the `termination_by`, it seems like And.decidable is not computable:
```
fail to show termination for
  decidable
with errors
failed to compile definition, consider marking it as 'noncomputable' because it depends on 'And.decidable', and it does not have executable code

failed to compile definition, consider marking it as 'noncomputable' because it depends on 'And.decidable', and it does not have executable code
```

Reply:
I missed your comment earlier. Your attempt works, but there were some issues in `Init.Logic`. After fixing those, this works:
```lean
instance All.decidable {P : α → Prop} : (t : Ordnode α) → [DecidablePred P] → Decidable (All P t)
| nil => decidableTrue
| node _ l _ r =>
  have : Decidable (All P l) := All.decidable l
  have : Decidable (All P r) := All.decidable r
  And.decidable
```
I pushed some commits. Can you fix the other decidability instances?

Reply:
I previously pushed fixes, but this pattern is better because it uses match & doesn't require noncomputable....